.dependency-cruiser.js

```diff
@@ -183,7 +238,140 @@ module.exports = {
           'npm-peer'
         ]
       }
-    }
+    },
+
+    /* application-specific architecture rules */
+    {
+      name: 'dependencies-of-shared',
+      severity: 'error',
+      from: {
+        path: dir('src/shared')
+      },
+      to: {
+        pathNot: [
+          dir('src/shared'),
+          dir('src/fs'),
+          ...npmDeps(
+            'node-fetch',
+            'markdown-escape',
+            'bn.js',
+          )
+        ]
+      }
+    },
+    {
+      name: 'dependencies-of-fs',
+      severity: 'error',
+      from: {
+        path: dir('src/fs')
+      },
+      to: {
+        pathNot: [
+          // TODO(SRe): is this okay?
+          dir('src/shared'), // 'shared' and 'fs' form a cycle
+          dir('src/fs'),
+          ...modules(
+            'path',
+            'fs',
+            'util',
+          ),
+        ]
+      }
+    },
+    {
+      name: 'dependencies-of-client',
+      severity: 'error',
+      from: {
+        path: dir('src/client'),
+        pathNot: [
+          file('src/client/webpack.config.js'),
+          testFilesWithinDir('src/client'),
+        ],
+      },
+      to: {
+        pathNot: [
+          dir('src/client'),
+          dir('src/shared'),
+          file('data/libraries/libraries.json'),
+          ...npmDeps(
+            '@fortawesome/fontawesome-free',
+            'clsx',
+            'easymde',
+            'query-string',
+            'react',
+            'react-alert',
+            'react-alert-template-basic',
+            'react-dom',
+            'react-loading-animation',
+            'react-portal-tooltip',
+            'react-responsive-modal',
+            'react-simplemde-editor',
+            'react-split-pane',
+            'remarkable',
+            'remarkable-react',
+            'scroll-into-view-if-needed',
+            'unicodeit',
+            'utf8',
+          ),
+        ]
+      }
+    },
+    {
+      name: 'dependencies-of-server',
+      severity: 'error',
+      from: {
+        path: dir('src/server'),
+        pathNot: [
+          dir('src/server/static'),
+          file('src/server/webpack.config.js'),
+          file('src/server/functions/webpack.config.js'),
+        ],
+      },
+      to: {
+        pathNot: [
+          dir('src/server'),
+          dir('src/shared'),
+          dir('src/fs'),
+          ...modules(
+            'child_process',
+            'fs',
+            'path',
+            'stream',
+            'url',
+          ),
+          ...npmDeps(
+            '@types',
+            'express',
+            'http-proxy-middleware',
+            'nodemailer',
+            'remarkable',
+          ),
+        ]
+      }
+    },
+    {
+      name: 'dependencies-of-vscode',
+      severity: 'error',
+      from: {
+        path: dir('src/vscode'),
+      },
+      to: {
+        pathNot: [
+          dir('src/vscode'),
+          dir('src/shared'),
+          dir('src/fs'),
```

Comment:
Just a remark: Removing this is a TODO item, and it turns out that this will be necessary to make LiveShare work properly.

Reply:
I've added a TODO comment for this.